contrib/babelfishpg_tsql/src/hooks.c

```
@@ -866,6 +905,58 @@ pltsql_ExecutorStart(QueryDesc *queryDesc, int eflags)
 			queryDesc->instrument_options |= INSTRUMENT_WAL;
 	}
 
+	/*
+	 * In TSQL dialect the RTE permissions might need to be checked against login mapped to given checkAsUser,
+	 * if it is valid, otherwise permissions are checked against session user (current login).
+	 */
+	if (sql_dialect == SQL_DIALECT_TSQL && queryDesc->plannedstmt != NULL)
+	{
+		ListCell	*lc;
+
+		foreach(lc, queryDesc->plannedstmt->permInfos)
+		{
+			RTEPermissionInfo	*perminfo = lfirst_node(RTEPermissionInfo, lc);
+			Oid             	relOid = perminfo->relid;
+
+			if (OidIsValid(relOid))
+			{
+				Oid schema_id = get_rel_namespace(relOid);
+
+				if (OidIsValid(schema_id))
+				{
+					char *nspname = get_namespace_name(schema_id);
+
+					/*
+					 * Check if relation's schema is valid and is not a shared schema. If yes,
+					 * then replace checkAsUser to its mapped login if present otherwise replace
+					 * with session user (current login).
+					 * We do not blindly want to check the permissions against session user (current login)
+					 * since permissions of RTEs inside a view are checked against that view's owner
+					 * which can very well be a user of some different database. So if we blindly check
+					 * permission against session user instead of view's owner thenit would break view's
+					 * ownership chaining. Instead, we will replace checkAsUser with it's corresponding mapped
+					 * login if present and only in cases where checkAsUser is not set, we will replace it
+					 * with session user (login). We are using login to allow cross database queries since login
+					 * can access all its objects across the databases.
+					 */
+					if (nspname != NULL && !is_shared_schema(nspname))
+					{
+						if (OidIsValid(perminfo->checkAsUser))
+						{
+							Oid loginId = get_login_for_user(perminfo->checkAsUser, nspname);
+							if (OidIsValid(loginId))
```

In case of invalid loginId I think we should throw an error like "could not find mapped login for db: %s"

---

As mentioned in comment above, we will fallback to existing behavior and not change `checkAsUser` instead of raising error since in this way we will atleast be allowing non-cross db accesses for guest owned objects.

---

Can there be any scenario now, and in future where the current user we fall back to, can have access and does not throw error?

---

But in case the current user have permission then it should be able to access the object so it is seems intuitive to not throw error.